File: sharedComposables/src/commonMain/kotlin/dev/mateusz1913/f1results/composable/common/DriverRow.kt
package dev.mateusz1913.f1results.composable.common

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Icon
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.mateusz1913.f1results.datasource.data.qualifying_results.QualifyingResultType
import dev.mateusz1913.f1results.datasource.data.race_results.FastestLapType
import dev.mateusz1913.f1results.datasource.data.race_results.ResultType

@Composable
fun DriverQualifyingRow(result: QualifyingResultType, noDriverInfo: Boolean = false) {
    DriverRow(
        position = result.position,
        driverName = "${result.driver.givenName} ${result.driver.familyName}",
        driverNumber = result.number,
        driverConstructorName = result.constructor.name,
        driverTime = result.Q3 ?: result.Q2 ?: result.Q1,
        noDriverPoints = true,
        noDriverInfo = noDriverInfo
    )
}

@Composable
fun DriverResultRow(result: ResultType, noDriverInfo: Boolean = false) {
    DriverRow(
        position = result.positionText,
        driverName = "${result.driver.givenName} ${result.driver.familyName}",
        driverNumber = result.number,
        driverConstructorName = result.constructor.name,
        driverTime = result.time?.time ?: result.status,
        noDriverPoints = false,
        driverPoints = result.points,
        driverFastestLap = result.fastestLap,
        noDriverInfo = noDriverInfo
    )
}

@Composable
fun DriverRow(
    position: String,
    driverName: String,
    driverNumber: String? = null,
    driverConstructorName: String,
    driverTime: String,
    noDriverPoints: Boolean = false,
    driverPoints: String? = null,
    driverFastestLap: FastestLapType? = null,
    noDriverInfo: Boolean = false
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Row(
            modifier = Modifier.weight(1f),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            StandingPosition(position)
            Column(modifier = Modifier.weight(1f)) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    if (!noDriverInfo) {
                        Text(
                            driverName,
                            fontSize = 18.sp,
                            fontWeight = FontWeight.Bold,
                            modifier = Modifier.padding(horizontal = 10.dp, vertical = 2.dp)
                        )
                        Box(modifier = Modifier.padding(horizontal = 10.dp, vertical = 2.dp)) {
                            Column(
                                modifier = Modifier
                                    .border(
                                        1.dp,
                                        MaterialTheme.colors.onBackground,
                                        RoundedCornerShape(13.dp)
                                    )
                                    .size(26.dp),
                                horizontalAlignment = Alignment.CenterHorizontally,
                                verticalArrangement = Arrangement.Center
                            ) {
                                if (driverNumber != null) {
                                    Text(
                                        driverNumber,
                                        fontSize = 14.sp,
                                        fontWeight = FontWeight.Medium
                                    )
                                }
                            }
                        }
                    }
                }
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        driverConstructorName,
                        fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold,
                        modifier = Modifier.padding(horizontal = 10.dp, vertical = 2.dp)
                    )
                    Box(modifier = Modifier.padding(vertical = 2.dp)) {
                        Text(
                            driverTime,
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Medium,
                            modifier = Modifier.padding(end = 10.dp),
                            textAlign = TextAlign.Right
                        )
                    }
                }
            }
        }
        if (!noDriverPoints) {
            Row(
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Icon(
                    Icons.Filled.LocalFireDepartment,
                    contentDescription = null,
                    tint = MaterialTheme.colors.secondary.copy(alpha = if (driverFastestLap?.rank == "1") 1f else 0f)
                )
                Column(
                    modifier = Modifier.size(30.dp)
                        .clip(RoundedCornerShape(15.dp))
                        .padding(start = 2.dp, end = 10.dp, top = 2.dp, bottom = 2.dp),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.Center
                ) {
                    Text(driverPoints ?: "-", fontSize = 14.sp, fontWeight = FontWeight.Bold)
                }
            }
        }
    }
}

File: swiftApp/Shared/Presentation/Common/DriverRow.swift
import SwiftUI
import shared

struct DriverQualifyingRow: View {
    var result: QualifyingResultType
    var noDriverInfo: Bool = false
    
    var body: some View {
        DriverRow(position: result.position, noDriverInfo: noDriverInfo, driverName: "\(result.driver.givenName) \(result.driver.familyName)", driverNumber: result.number, driverConstructorName: result.constructor.name, driverTime: result.Q3 ?? result.Q2 ?? result.Q1, noDriverPoints: true, driverPoints: nil, driverFastestLap: nil)
    }
}

struct DriverResultRow: View {
    var result: ResultType
    var noDriverInfo: Bool = false
    
    var body: some View {
        DriverRow(position: result.positionText, noDriverInfo: noDriverInfo, driverName: "\(result.driver.givenName) \(result.driver.familyName)", driverNumber: result.number, driverConstructorName: result.constructor.name, driverTime: result.time?.time ?? result.status, noDriverPoints: false, driverPoints: result.points, driverFastestLap: result.fastestLap)
    }
}

struct DriverRow: View {
    var position: String
    var noDriverInfo: Bool = false
    var driverName: String
    var driverNumber: String?
    var driverConstructorName: String
    var driverTime: String
    var noDriverPoints: Bool = false
    var driverPoints: String?
    var driverFastestLap: FastestLapType?
    
    var body: some View {
        HStack(alignment: .center) {
            HStack(alignment: .center) {
                StandingPosition(position: position)
                VStack(alignment: .center) {
                    HStack(alignment: .center) {
                        if !noDriverInfo {
                            Text(driverName)
                                .font(.system(size: 18))
                                .fontWeight(.bold)
                                .frame(minWidth: 0, maxWidth: .infinity, alignment: .leading)
                                .padding(.horizontal, 10)
                                .padding(.vertical, 2)
                            VStack {
                                VStack(alignment: .center) {
                                    if let number = driverNumber {
                                        Text(number)
                                            .font(.system(size: 14))
                                            .fontWeight(.medium)
                                    }
                                }
                                .size(26)
                                .overlay(RoundedRectangle(cornerRadius: 13).stroke(lineWidth: 1))
                            }
                            .padding(.horizontal, 10)
                            .padding(.vertical, 2)
                        }
                    }
                    .fillMaxWidth()
                    HStack(alignment: .center) {
                        Text(driverConstructorName)
                            .font(.system(size: 16))
                            .fontWeight(.bold)
                            .frame(minWidth: 0, maxWidth: .infinity, alignment: .leading)
                            .padding(.horizontal, 10)
                            .padding(.vertical, 2)
                        VStack {
                            Text(driverTime)
                                .font(.system(size: 14))
                                .fontWeight(.medium)
                                .padding(.trailing, 10)
                        }
                        .padding(.vertical, 2)
                    }
                    .fillMaxWidth()
                }
                .fillMaxSize()
            }
            .fillMaxSize()
            if !noDriverPoints {
                HStack(alignment: .center, spacing: 0) {
#if os(iOS)
                    if let uiImage = UIImage(named: "baseline_local_fire_department_black_24pt") {
                        Image(uiImage: uiImage)
                            .foregroundColor(Color.secondaryColor.opacity(driverFastestLap?.rank == "1" ? 1 : 0))
                    }
#elseif os(macOS)
                    if let nsImage = NSImage(named: "baseline_local_fire_department_black_24pt") {
                        Image(nsImage: nsImage)
                            .foregroundColor(Color.secondaryColor.opacity(driverFastestLap?.rank == "1" ? 1 : 0))
                    }
#endif
                    ZStack(alignment: .center) {
                        Text(driverPoints ?? "-")
                            .font(.system(size: 14))
                            .fontWeight(.bold)
                    }
                    .size(30)
                    .clipShape(Circle())
                }
                .padding(.init(top: 2, leading: 2, bottom: 2, trailing: 10))
            }
        }
        .fillMaxWidth()
    }
}

struct DriverQualifyingRow_Preview: PreviewProvider {
    static var previews: some View {
        Group {
            DriverQualifyingRow(result: QualifyingResultType(number: "3", position: "7", driver: DriverType(driverId: "ricciardo", permanentNumber: "3", code: "RIC", url: nil, givenName: "Daniel", familyName: "Ricciardo", dateOfBirth: "01-07-1989", nationality: "Australia"), constructor: ConstructorType(constructorId: "mclaren", url: nil, name: "McLaren", nationality: "United Kingdom"), Q1: "1:23.456", Q2: "1:22.333", Q3: "1:21.333"))
            DriverQualifyingRow(result: QualifyingResultType(number: "3", position: "7", driver: DriverType(driverId: "ricciardo", permanentNumber: "3", code: "RIC", url: nil, givenName: "Daniel", familyName: "Ricciardo", dateOfBirth: "01-07-1989", nationality: "Australia"), constructor: ConstructorType(constructorId: "mclaren", url: nil, name: "McLaren", nationality: "United Kingdom"), Q1: "1:23.456", Q2: "1:22.333", Q3: "1:21.333"), noDriverInfo: true)
        }
        .previewLayout(PreviewLayout.fixed(width: 400, height: 100))
    }
}

struct DriverResultRow_Preview: PreviewProvider {
    static var previews: some View {
        Group {
            DriverResultRow(result: ResultType(number: "3", position: "6", positionText: "6", points: "8", driver: DriverType(driverId: "ricciardo", permanentNumber: "3", code: "RIC", url: nil, givenName: "Daniel", familyName: "Ricciardo", dateOfBirth: "01-07-1989", nationality: "Australia"), constructor: ConstructorType(constructorId: "mclaren", url: nil, name: "McLaren", nationality: "United Kingdom"), grid: "7", laps: "65", status: "", time: DurationType(millis: nil, time: "+12.0"), fastestLap: FastestLapType(rank: "3", lap: "4", time: DurationType(millis: "123", time: "1:23"), averageSpeed: AverageSpeedType(speed: "260", units: "kmh"))))
            DriverResultRow(result: ResultType(number: "3", position: "6", positionText: "6", points: "8", driver: DriverType(driverId: "ricciardo", permanentNumber: "3", code: "RIC", url: nil, givenName: "Daniel", familyName: "Ricciardo", dateOfBirth: "01-07-1989", nationality: "Australia"), constructor: ConstructorType(constructorId: "mclaren", url: nil, name: "McLaren", nationality: "United Kingdom"), grid: "7", laps: "65", status: "", time: DurationType(millis: nil, time: "+12.0"), fastestLap: FastestLapType(rank: "1", lap: "4", time: DurationType(millis: "123", time: "1:23"), averageSpeed: AverageSpeedType(speed: "260", units: "kmh"))), noDriverInfo: true)
        }
        .previewLayout(PreviewLayout.fixed(width: 400, height: 100))
    }
}
